flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -329,37 +330,50 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
         // materialization may truncate only a part of the previous result and the backend would
         // have to split it somehow for the former option, so the latter is used.
         lastCheckpointId = checkpointId;
-        lastUploadedFrom = materializedTo;
+        lastUploadedFrom = periodicMaterializer.lastMaterializedTo();
         lastUploadedTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
 
         LOG.debug(
                 "snapshot for checkpoint {}, change range: {}..{}",
                 checkpointId,
                 lastUploadedFrom,
                 lastUploadedTo);
+
+        SnapshotResult<KeyedStateHandle> asOfNow = periodicMaterializer.materializedSnapshot();
+
         return toRunnableFuture(
                 stateChangelogWriter
                         .persist(lastUploadedFrom)
-                        .thenApply(this::buildSnapshotResult));
+                        .thenApply(delta -> buildSnapshotResult(delta, asOfNow)));
     }
 
-    private SnapshotResult<KeyedStateHandle> buildSnapshotResult(ChangelogStateHandle delta) {
-        // Can be called by either task thread during the sync checkpoint phase (if persist future
-        // was already completed); or by the writer thread otherwise. So need to synchronize.
-        // todo: revisit after FLINK-21357 - use mailbox action?
-        synchronized (materialized) {
-            // collections don't change once started and handles are immutable
-            List<ChangelogStateHandle> prevDeltaCopy = new ArrayList<>(restoredNonMaterialized);
-            if (delta != null && delta.getStateSize() > 0) {
-                prevDeltaCopy.add(delta);
-            }
-            if (prevDeltaCopy.isEmpty() && materialized.isEmpty()) {
-                return SnapshotResult.empty();
-            } else {
-                return SnapshotResult.of(
-                        new ChangelogStateBackendHandleImpl(
-                                materialized, prevDeltaCopy, getKeyGroupRange()));
-            }
+    @Override
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        periodicMaterializer.triggerMaterialization();
+    }
+
+    private SnapshotResult<KeyedStateHandle> buildSnapshotResult(
+            ChangelogStateHandle delta,
+            SnapshotResult<KeyedStateHandle> materialized) {
+
+        // collections don't change once started and handles are immutable
+        List<ChangelogStateHandle> prevDeltaCopy =
+                new ArrayList<>(periodicMaterializer.restoredNonMaterialized());
```

Comment:
I'm afraid this access needs synchronization:
- `periodicMaterializer.restoredNonMaterialized` is written by the task thread
- this method `buildSnapshotResult()` can be called by the DFS writer thread